### PR TITLE
Add 'endpoints' endpoint

### PR DESCRIPTION
**Pull request by nickbp**
Common replacement for framework-specific `/connections` endpoints and CLI commands currently implemented by Cassandra/Kafka/HDFS.

The built-in `TaskInfo` search relies on the presence of `DiscoveryInfo`s containing the relevant port information (something that `hello-world` currently doesn't do -- as its tasks don't serve any requests). If the `DiscoveryInfo` has VIP label information, that is automatically included as well, otherwise the endpoints returned will just be based on the information in the base `DiscoveryInfo`.

In addition to the built-in `DiscoveryInfo` lookup, framework authors may also provide their own custom endpoint content (e.g. Kafka's zk path, or HDFS' xml files).

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            JSONObject vipEndpoint = endpointsByName.get(vipInfo.name);
+            if (vipEndpoint == null) {
+                vipEndpoint = new JSONObject();
+                endpointsByName.put(vipInfo.name, vipEndpoint);
+            }
+
+            // append entry to 'direct' array for this task:
+            vipEndpoint.append(RESPONSE_KEY_DIRECT, directHostPort);
+            // populate 'vip' field if not yet populated (due to another task with the same vip):
+            vipEndpoint.put(RESPONSE_KEY_VIP, String.format("%s.%s.l4lb.thisdcos.directory:%d",
```
Comment (mrbrowning) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java]: Should probably extract the VIP URL format to a constant somewhere for visibility.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+     * This may be used to expose additional endpoint types independently of what's listed in
+     * {@link DiscoveryInfo}, such as a Kafka service exposing a Zookeeper path that's separate from
+     * the default broker host/port listing.
+     *
+     * This only supports simple string values in order to ensure that the 'endpoints' endpoint
+     * remains relatively consistent across services. For a per-task listing of endpoints, you
+     * should provide that information via {@link DiscoveryInfo} in your {@link TaskInfo} and they
+     * will appear automatically.
+     *
+     * @param valueProducer the endpoint producer, which will be invoked whenever a user queries the
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java]: Done

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java (hunk 1 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/EndpointsResource.java]: Done

**Comment by nickbp** [comment]
retest this please

**Pull request merged**